### ISSUE TITLE
Saving cause segmentation fault

### ISSUE THREAD

**Issue by aapo** [raceintospace/raceintospace#138]
Start game: BARIS_SAVE=/home/user/.raceintospace BARIS_DATA=/opt/ris/data gdb ./raceintospace

 Press A, Press T (time capsule), Click save, give name, press enter -> Segmentation fault.

```
Program received signal SIGSEGV, Segmentation fault.
0x4037d88c in fread () from /lib/libc.so.6
(gdb) bt
#0  0x4037d88c in fread () from /lib/libc.so.6
#1  0x00042de8 in FileAccess(char) ()
#2  0x0004078c in Admin(char) ()
#3  0x000f4758 in PortSel(char, char) ()
#4  0x000f3928 in Port(char) ()
#5  0x000f119c in Master(char) ()
#6  0x000a1d14 in MainLoop() ()
#7  0x000a0ad4 in game_main_impl(int, char**) ()
#8  0x000a0e5c in game_main ()
#9  0x00136e48 in main ()
```

**Comment by rnyoakum** [comment]
This issue was introduced in commit eb75c92c3ca0a775f6aed332d270ba31dc623393. I should be able to fix it quickly enough, but the save code is pretty messy as a result of the whole normal / modem / play be email situation. As a result, I'm not sure what information is supposed to be included in a normal save. Is a save supposed to include only events up to the beginning of the turn, or should it include information representing the game state at the moment of the save (future missions planned, research performed, etc.)?

**Comment by rnyoakum** [comment]
In commit eb75c92c3ca0a775f6aed332d270ba31dc623393, an error was introduced that affected writing the event data (the newscast text) to a save file. This information is stored so it can be reviewed on later turns at the Viewing Stand. The error caused the game, rather than writing out the event data, to try reading from an uninitialized file pointer. This would likely generate a segmentation fault and the OS would shut down the game. This error should be fixed by commit 8c98fa5181d742f8a957995fb76723104c8f50a3.

The second issue is that, since the introduction of the error, normal save files have not been written properly. Autosaves, using different code, managed to avoid this problem (but may have their own issues). However, autosaves generated after loading a non-autosave file will inherit the corruption of the other file. Save files consist of four sections:

  I) A save file header
  II) The game state
  III) Mission replay data
  IV) Event data (the newscasts)

Because the save file header and game state (i.e. the Data variable) are both written correctly, the game did not register that the save files are corrupt. It should have, and I intend to upgrade the error checking so the game will pick up on this type of problem in the future. The game, nonetheless, proceeds to try and load the event data, ignoring that it isn't reading anything.

I've also observed that a consequence was that normal saves, at least in my case, are all the same length. The save file header is a set size, which should not change, but the game state data is compressed before being written, so I suspect it should be varying in size. This, along with some testing I've done, suggests the mission replay data is also not being fully written. I could be wrong, but if not I hypothesize part of the replay data is buffered for writing when the segmentation fault occurs and kills game IO before the buffer can flush.

The lack of event data and suspected corruption of the Replay data appears to be responsible for the Viewing Stand generating segmentation faults. This was an issue of which Leon and I were aware, but had not reported because I could not reliably replicate it in new games. The Viewing stand still has problems with the event data, but that demands its own fix.

Overall, the save/load system needs a lot of attention, and I've been working on an overhaul that would handle some...legacy... issues. As a part of this, any saves (excluding autosaves) created in the past four years or so might be declared corrupted and fail to load. If this is a major issue for anyone, I can see about either adding an Allow Unsafe Saves config option or creating a utility to make corrupt saves "safe."

If anyone has a proposal on handling this long-term save corruption issue, please let me know.

**Comment by peyre** [comment]
I appreciate the attention you're giving to the save system, Ryan!  As far as I'm concerned at least, losing games we've saved in the past several years isn't an issue; I'd say the priority is to fix the save system even if we lose whatever saves we've made previously.  That'll solve the problem going forward.
